### DISCUSSION TITLE
Exclude workloads from scanning

### DISCUSSION THREAD

**Discussion by kfox1111** [aquasecurity/starboard#670] [Development]
Really not sure about this one. Probably should just have a discussion...

gitlab dynamically submits jobs to k8s. the jobs run for a short amount of time (seconds to a minute) usually and then disappear. Starboard-operator is doing scans on those jobs though and then the vulnerability report gets deleted when gitlab cleans up the job. This is causing the images to be double downloaded and cpu used to scan something that disappears quickly.

Trying to decide what do / recommend about these things. Some options:
* Delay submitting vulnerability jobs for a certain amount of time to see if they will go away.
* Support ignoring workloads based on certain workload labels (app=gitlab-job)
* Cache image scans so lots of gitlab job runs just reuse the data.
* Other?

What do you think?

**Comment by danielpacak** [comment]
If we cannot skip a namespace in which operator discovers (GitLab) jobs then we should probably provide a way to ignore certain workload kinds (e.g. jobs or cronjobs with frequent schedule) and / or respect configurable label selectors to exclude particular workloads from scanning.

/cc @amitbis

**Reply by kfox1111** [reply]
Often we have gitlab jobs deploying services into the same namespace. so the workloads will be mixed in a single namespace.

I think gitlab-runner uses deployments for its jobs unfortunately. So labels on the workload may be the best we can do there to filter them.

**Comment by danielpacak** [comment]
Another requirement I saw was for excluding namespaces in #638 . This brings me to the conclusion that maybe we should define exclusion semantics at different levels:
- Namespaces - for example, exclude namespace foo, bar, and baz
- K8s resource kinds - for example, exclude Jobs, CronJobs, DaemonSets, and ConfigMaps
- K8s resource instances - for example, exclude pods, ReplicaSets or ConfigMaps that match certain label, annotation or field selectors

**Reply by MPritsch** [reply]
Excluding certain namespaces or pods by label would help us with keeping our cluster maintainable after being hit by this bug with secret creation for certain pods https://github.com/aquasecurity/starboard/issues/1038
While this issue might be resolved soon, I have to think that an exclude functionality could help as a workaround for future issues.
